### DISCUSSION TITLE
Implement json_patch for sqlite

### DISCUSSION THREAD

**Discussion by Sytten** [diesel-rs/diesel#3954] [Q&A] [answered]
Hi!

I am trying to play around with custom functions.

```rust
diesel::table! {
    requests (id) {
        id -> Integer,
        metadata -> Nullable<Text>,
    }
}

sql_function! { fn json_patch(t: Text, p: Text) -> Text; }

diesel::update(requests::table.filter(requests::id.eq(1)))
                            .set(requests::metadata.eq(json_patch("{}", "{}")))
                            .returning(requests::metadata)
                            .get_result::<RequestMetadata>(conn)?;
```

This errors out since the types don't match:
```
error[E0271]: type mismatch resolving `<json_patch<Bound<Text, &str>, Bound<Text, &str>> as Expression>::SqlType == Nullable<Text>`
  --> xxxxxxxx.rs:29:53
   |
29 | ...                   .set(requests::metadata.eq(json_patch("{}", "{}")))
   |                                               ^^ expected `Text`, found `Nullable<Text>`
   |
   = note: expected struct `diesel::sql_types::Text`
              found struct `diesel::sql_types::Nullable<diesel::sql_types::Text>`
   = note: required for `json_patch<Bound<Text, &str>, Bound<Text, &str>>` to implement `diesel::expression::AsExpression<diesel::sql_types::Nullable<diesel::sql_types::Text>>`
```

What is the preferred way to fix that? It's easy to change to `sql_function! { fn json_patch(t: Text, p: Text) -> Nullable<Text>; }`, but then it makes unusable for columns that are not nullable. `json_patch` only returns null if one of the parameters is null AFAIK.

**Comment by weiznich** [comment, marked as answer]
If the function is always guaranteed to not return `NULL` it's correct to write that this way. The correct way to solve this currently is to call [`NullableExpressionMethods::nullable()`](https://docs.diesel.rs/2.1.x/diesel/expression_methods/trait.NullableExpressionMethods.html#method.nullable) on the result afterwards:

```rust

diesel::update(requests::table.filter(requests::id.eq(1)))
                            .set(requests::metadata.eq(json_patch("{}", "{}").nullable()))
                            .returning(requests::metadata)
                            .get_result::<RequestMetadata>(conn)?;
```

**Reply by Sytten** [reply]
Thanks, I am sure this was answered somewhere but GitHub search isnt the best. I will try to add a more complex example to the documentation, it works really well the macro but a bit daunting to use.

**Reply by weiznich** [reply]
The "complicated" part here is that you could combine nearly everything that's in some of the `*ExpressionMethods` traits with anything that implements `Expression`. So it's likely not just a matter of adding some cross references to the documentation, but rather writing a full guide for how these expressions are expected to work. 

(Adding these cross references might be helpful as well)
